crates/xline-client/src/clients/lock.rs

```diff
@@ -1,78 +1,242 @@
+            let (mut keeper, mut stream) = lease_client
+                .keep_alive(LeaseKeepAliveRequest::new(lease_id))
+                .await?;
+            loop {
+                keeper.keep_alive()?;
+                if let Some(resp) = stream.message().await? {
+                    if resp.ttl < 0 {
+                        return Err(XlineClientError::InvalidArgs(String::from(
+                            "lease keepalive response has negative ttl",
+                        )));
+                    }
+                    sleep(Duration::from_secs(resp.ttl.unsigned_abs() / 3)).await;
+                }
+            }
+        }));
+        Ok(self)
     }
 }
 
-// These methods primarily originate from xline lock server,
-// see also: `xline/src/server/lock_server.rs`
-impl LockClient {
-    /// Creates a new `LockClient`
+impl Drop for Session {
     #[inline]
-    pub fn new(
-        curp_client: Arc<CurpClient>,
-        channel: Channel,
-        token: Option<String>,
-        id_gen: Arc<LeaseIdGenerator>,
-    ) -> Self {
+    fn drop(&mut self) {
+        if let Some(keep_alive) = self.keep_alive.take() {
+            keep_alive.abort();
+        }
+    }
+}
+
+/// Xutex（Xline Mutex） implements the sync lock with xline
+#[derive(Debug)]
+pub struct Xutex {
+    /// Lock session
+    session: Session,
+    /// Lock
+    prefix: String,
+    /// Lock key
+    key: String,
+    /// The revision of lock key
+    rev: i64,
+    /// Request header
+    header: Option<ResponseHeader>,
+}
+
+/// An RAII implementation of `Xutex`
+#[derive(Default, Debug)]
+pub struct XutexGuard {
+    /// The lock client that used to unlock `key` when `XutexGuard` is dropped
+    client: Option<LockClient>,
+    /// The key that the lock held
+    key: String,
+}
+
+impl XutexGuard {
+    /// Create a new `XutexGuard`
+    fn new(client: LockClient, key: String) -> AsyncDropper<Self> {
+        AsyncDropper::new(Self {
+            client: Some(client),
+            key,
+        })
+    }
+}
+
+#[async_trait]
+impl AsyncDrop for XutexGuard {
+    #[inline]
+    async fn async_drop(&mut self) {
+        if let Some(ref client) = self.client {
+            let _ignore = client.delete_key(self.key.as_bytes()).await;
+        }
+    }
+}
+
+impl Xutex {
+    /// Create an Xutex
+    #[inline]
+    #[must_use]
+    pub fn new(session: Session, prefix: &str) -> Self {
         Self {
-            curp_client: Arc::clone(&curp_client),
-            lease_client: LeaseClient::new(curp_client, channel.clone(), token.clone(), id_gen),
-            watch_client: WatchClient::new(channel, token.clone()),
-            token,
+            session,
+            prefix: format!("{prefix}/"),
+            key: String::new(),
+            rev: -1,
+            header: None,
         }
     }
 
+    /// Get the key of the Xutex
+    #[inline]
+    #[must_use]
+    pub fn key(&self) -> &str {
+        self.key.as_str()
+    }
```

Comment:
Should this method be moved to `XutexGuard`? Because the key only usable after the mutex has been locked.

Reply:
Done